Exclusion of source fields

scampi opened the pull request:
This PR adds filtering capability of source fields for a given index pattern. Excluded fields are then not retrieved by a saved search. However, looking at a document using the `doc` API displays all available fields.

In the index settings, a new column `excluded` is added which indicates if a field is excluded or not. In the picture, the `born` field is excluded. There is also a new tab `source filters` which shows all the exclusion patterns.

![2016-10-06-232257_1200x729_scrot](https://cloud.githubusercontent.com/assets/795879/19172894/3b84c94a-8c1c-11e6-9818-971ed6c77521.png)

Below is the `source filters` tab which shows on the left side the filter pattern, as well as the fields that match.

![2016-10-06-232312_1200x729_scrot](https://cloud.githubusercontent.com/assets/795879/19172893/3b8211fa-8c1c-11e6-996e-cccf44a19e6e.png)

On the discover page, I had created a saved search showing the field `born`. It is now empty since it is excluded.

![2016-10-06-232334_1200x729_scrot](https://cloud.githubusercontent.com/assets/795879/19172892/3b804b5e-8c1c-11e6-9f83-762c0ecf7766.png)

However, if you look at a specific document, all available fields are displayed.

![2016-10-06-232352_1200x729_scrot](https://cloud.githubusercontent.com/assets/795879/19172891/3b803b64-8c1c-11e6-9f2e-db1b2231da64.png)

close #4366.
Reimplementation of #5981.

elasticsearch-release commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

scampi commented:
Errors in jenkins seem unrelated to this PR. I see many of the following:

```
TypeError: _this.saveScreenshot(...).finally is not a function
  at Command.<anonymous>  <test/support/pages/common.js:259:17>
  at runCallback  <node_modules/intern/node_modules/leadfoot/Command.js:526:31>
  at Command.<anonymous>  <node_modules/intern/node_modules/leadfoot/Command.js:545:11>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at run  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at <node_modules/intern/node_modules/dojo/nextTick.ts:44:3>
  at nextTickCallbackWith0Args  <node.js:420:9>
  at process._tickCallback  <node.js:349:13>
  at Command.then  <node_modules/intern/node_modules/leadfoot/Command.js:542:10>
  at Command.catch  <node_modules/intern/node_modules/leadfoot/Command.js:556:15>
  at Test.test  <test/functional/apps/discover/_shared_links.js:132:17>
  at <node_modules/intern/lib/Test.js:211:24>
  at <node_modules/intern/node_modules/dojo/Promise.ts:393:15>
  at runCallbacks  <node_modules/intern/node_modules/dojo/Promise.ts:11:11>
  at <node_modules/intern/node_modules/dojo/Promise.ts:317:4>
  at run  <node_modules/intern/node_modules/dojo/Promise.ts:237:7>
  at <node_modules/intern/node_modules/dojo/nextTick.ts:44:3>
  at nextTickCallbackWith0Args  <node.js:420:9>
  at process._tickCallback  <node.js:349:13>
```

scampi commented:
seems related to https://github.com/elastic/kibana/issues/7327#issuecomment-224859084

LeeDr commented:
The first failure screenshot shows this;
![failure_1465461668876_query_save query should show toast message and display query name](https://cloud.githubusercontent.com/assets/13542669/15930756/376c34b6-2e1a-11e6-864b-16ddd4cb7425.png)

The test failed to find the toast message "Discover: Saved Data Source "Query # 1" even though it is there, because there's another toast error message that says "! Discover: this.fieldFilters is undefined".

I haven't figured out why that message is being displayed.

scampi commented:
Oh thanks @LeeDr ! I didn't pay attention to that. I will check that error.

scampi commented:
I added the missing `fieldFilters` field in the test dumps `test/fixtures/dump_data/`.

epixa commented:
jenkins, test it

scampi commented:
@epixa I have merged with master, as the refactoring of the settings section impacted on what is proposed in this PR. Please have a look if it is all fine.
The change to `src/plugins/kibana/server/lib/schemas/resources/index_pattern_schema.js` is needed when you create the index pattern through the CSV upload mechanism.

Bargs commented on src/plugins/kibana/server/lib/schemas/resources/index_pattern_schema.js:
```diff
@@ -6,6 +6,7 @@ module.exports = Joi.object({
   time_field_name: Joi.string(),
   interval_name: Joi.string(),
   not_expandable: Joi.boolean(),
+  fieldFilters: Joi.string().default('[]'),
```
To stay consistent with the way the other optional fields are handled, I don't think we should set a default here. An index pattern could be created in a number of ways (including a user creating it directly in ES) so we can't rely on this default anyway.

epixa replied:
This should probably use underscores rather than camel case as well, to be consistent with the rest of the properties.

scampi replied:
The mapping of the index pattern in https://github.com/elastic/kibana/blob/91b6d925292f7bb62f059f2ba06c41e775ef2eb7/src/ui/public/index_patterns/_index_pattern.js#L41 is camel case. Out of curiosity, is it normal ? Or will it change to reflect the snake_case format ?

Bargs replied:
The code here (in the PR) defines a schema used by a new Kibana API. It's actually independent of the format that's ultimately stored in Elasticsearch.

The code you've linked to is older and interacts directly with Elasticsearch, which is why it matches exactly what you see in ES.

So the short answer is yes, `index_pattern_schema.js` should be snake_cased while `_index_pattern.js` should be camelCased.

scampi replied:
Thanks for the clarification on that!

scampi replied:
BTW, where does the translation from the new Kibana API to the format stored in Elasticsearch take place ?

Bargs commented:
@scampi is this ready for a thorough review, or are you still working on it?

scampi commented:
@Bargs I will just address the last two comments raised by you and court, and it will then be good for a thorough review! I didn't have time recently to do that.

elasticmachine commented:
Can one of the admins verify this patch?

scampi commented:
@Bargs this PR is now ready to have a thorough review.
I removed the default from the index pattern schema.

Bargs commented:
Hi @scampi, sorry I've taken so long to get around to this. Could you merge your branch with the latest from master? Some directories have changed around a bit. I'll get on the review as soon as it's up to date.

elasticmachine commented:
Can one of the admins verify this patch?

scampi commented:
No problem @Bargs , there you go.

Bargs commented:
Thanks @scampi. There are still some files that need re-arranged though. The `src/plugins` directory became `src/core_plugins`, so all the files under `src/plugins` need moved and the references to them need to be updated. Sorry for the trouble!

scampi commented:
I had not seen that plugins folder was renamed, thanks.
I updated the `getComputedFields` method to get the exclusion pattern set for an index. However, this means that the doc API http://localhost:5601/rvw/app/kibana#/doc/aaa/aaa/default?id=myindex:4&_g=() will show the filtered document. Is that OK ?

Bargs commented:
@scampi hmmm, good question. I think it's reasonable to hide them by default, since the fields could be quite large. How difficult do you think it would be to add a "Show all fields" link that could retrieve the full doc upon request? At the very least, we should probably add a message to the doc page that says something like "Some document fields are currently filtered out based on your index pattern configuration"

Bargs commented:
@scampi didn't the field filter input used to dynamically show you what fields matched, as you typed? Or am I dreaming things?

scampi commented:
hmm I don't think it was ever doing that @Bargs . However, @spalger was the original author of that particular part of the code.

Something that came to my mind: @spalger had changed that part of the code https://github.com/elastic/kibana/pull/7402/files#diff-333ab9232d101ea65556361cc90c60edR22 to apply the exclusion pattern. However, I think it is not needed anymore after the recent refactoring of the kibana code base, since the source is now set there https://github.com/elastic/kibana/pull/7402/files#diff-d1695ba2026ff89878f9e4f4de683758R50.
I am not that confident with that low-level code of kibana, so i'd love some advice here ;o)

Regarding the `doc` API, I will try to add a `show all` button and some message if case of excluded fields.

Bargs commented:
I'm not sure of the details there either :/

@spalger could you enlighten us as to what [this line](https://github.com/elastic/kibana/pull/7402/files#diff-d1695ba2026ff89878f9e4f4de683758R71) is doing, and how it interacts with [this use of computed fields](https://github.com/Bargs/kibana/blob/90894ea7aeda5a74a11c7b990e0f56ef23c99a83/src/ui/public/courier/data_source/_abstract.js#L290-L290) in the abstract data source?

Bargs commented:
@scampi If I remember correctly, you had a gist of all the todo items from the previous pull. Any chance you could link that? It's been so long I forget what all I said.

scampi commented:
This is the gist @Bargs  https://gist.github.com/scampi/30b8143c40179cdc07cbca1562ea59b3

Bargs commented:
@scampi thanks! Another question: how would you feel about renaming 'Field filters' to 'Source filters' in the UI? I think that might make things more clear. For instance, as I was testing I realized my wildcards weren't matching `raw` fields. I had to sit and think about it a bit and realize raw fields won't be retrieved since they're not a part of _source. I think this also clears up any confusion about scripted fields, meta fields, etc. not matching since they're also not a part of _source.

scampi commented:
@Bargs sure no problem. It was indeed open to misinterpretation.

scampi commented:
@Bargs about https://github.com/elastic/kibana/pull/7402#issuecomment-235618759, I was thinking that maybe we can explicitly retrieve all fields in this view, i.e., set `_source` to `true` at [doc.js:54](https://github.com/elastic/kibana/blob/81efe3b/src/core_plugins/kibana/public/doc/controllers/doc.js#L54).
It doesn't really make sense to filter out anything in this particular view it seems. What do you think ?

Bargs commented:
@scampi I'm cool with that.

scampi commented:
@Bargs I've renamed `Field Filters` to `Source Filters` and corrected the doc controller to explicitly retrieve the source.
I think only my question in https://github.com/elastic/kibana/pull/7402#issuecomment-235629055 is left for this PR ?

Bargs commented:
@scampi I'm not 100% sure what the application of source filtering in search_source.js is for, but I do know that `getComputedFields` is used [here](https://github.com/Bargs/kibana/blob/90894ea7aeda5a74a11c7b990e0f56ef23c99a83/src/ui/public/courier/data_source/_abstract.js#L290-L290) which determines the `_source` property on the actual request to Elasticsearch from Discover. So I think we do need to keep the change in `_get_computed_fields.js`

I'm start to review the rest of the code now as well

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/_edit.html:
```diff
@@ -32,22 +32,35 @@
+        ng-show="state.tab == 'sourceFilters'"
+        index-pattern="indexPattern"
+        class="fields field-filters">
+      </field-filters>
```
You can place the closing bracket and the closing tags on the same line, as described here https://github.com/elastic/kibana/blob/master/style_guides/html_style_guide.md

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/_edit.js:
```diff
@@ -65,8 +66,8 @@ uiModules.get('apps/management')
-  $scope.$watchCollection('indexPattern.fields', function () {
-    $scope.conflictFields = _.filter($scope.indexPattern.fields, {type: 'conflict'});
+  $scope.$watchCollection('indexPattern.fields', function (fields) {
+    $scope.conflictFields = _.filter(fields, { type: 'conflict' });
```
what was the reason for this change?

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/_field_types.js:
```diff
@@ -11,14 +11,21 @@ export default function GetFieldTypes() {
       scripted: 0
```
Now that we're adding source filters, this feels like a poorly named file. We should probably just rename it as _edit_sections.js or something like that.

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/_indexed_fields.js:
```diff
@@ -25,6 +27,7 @@ uiModules.get('apps/management')
         { title: 'format' },
         { title: 'analyzed', info: 'Analyzed fields may require extra memory to visualize' },
         { title: 'indexed', info: 'Fields that are not indexed are unavailable for search' },
+        { title: 'exclude', info: 'Fields that are excluded from _source when it is fetched' },
```
The title should probably be `excluded` instead of `exclude` to match `indexed` and `analyzed`.

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/field_filters/field_filters.html:
```diff
@@ -0,0 +1,97 @@
+<h3>Source Filters</h3>
```
Could you also rename the file itself to `source_filters.html`?

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/field_filters/field_filters.html:
```diff
@@ -0,0 +1,97 @@
+            name="form">
+
+            <div class="value">
+              <span ng-if="fieldFilters.editting !== filter">{{ filter.value }}</span>
```
Typo `editing`

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/field_filters/field_filters.html:
```diff
@@ -0,0 +1,97 @@
+  Source filters can be used to exclude one or more fields when fetching the document source. This happens when viewing a document in the Discover app, or with a table displaying results from a saved search in the Dashboard app. Each row is built using the source of a single document, and if you have documents with large or unimportant fields you may benefit from filtering those out at this lower ...
+</p>
+
+<div ng-class="{ saving: fieldFilters.saving }" class="field-filters-container">
```
I'd update `fieldFilters` to `sourceFilters` throughout this file as well (and in any other files where that name is used) for the sake of consistency.

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/field_filters/field_filters.html:
```diff
@@ -0,0 +1,97 @@
+                ng-model="filter.value"
+                input-focus
+                ng-if="fieldFilters.editting === filter"
+                placeholder="field name filter, accepts wildcards (e.g., `user*` to filter fields starting with 'user')"
```
~~field name filter~~ source filter

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/field_filters/field_filters.html:
```diff
@@ -0,0 +1,97 @@
+            <div class="input-group">
+              <input
+                ng-model="fieldFilters.newValue"
+                placeholder="field name filter, accepts wildcards (e.g., `user*` to filter fields starting with 'user')"
```
~~field name filter~~ source filter

Bargs commented:
There's a lot of wasted horizontal space on the new Source Filter tab. What do you think about moving to a two column layout, with the filter list/editor and known matches displayed side by side?

Bargs commented:
@scampi ignore my other comment above, you were right, the following lines can be deleted:

https://github.com/elastic/kibana/pull/7402/files#diff-333ab9232d101ea65556361cc90c60edR22
https://github.com/elastic/kibana/blob/master/src/ui/public/courier/data_source/_abstract.js#L288

When you make this change, can you please add a commit message that explains that the changes in https://github.com/elastic/kibana/pull/7402/files#diff-d1695ba2026ff89878f9e4f4de683758R50 make these two lines unnecessary?

Bargs commented on src/ui/public/courier/data_source/_abstract.js:
```diff
@@ -382,6 +384,11 @@ export default function SourceAbstractFactory(Private, Promise, PromiseEmitter)
+        if (flatState.body._source && flatState.body.fielddata_fields) {
+          const filter = fieldWildcardFilter(flatState.body._source.exclude);
+          flatState.body.fielddata_fields = flatState.body.fielddata_fields.filter(filter);
+        }
```
I'm pretty sure this block could be moved up to where `fielddata_fields` is initialized: https://github.com/elastic/kibana/blob/90894ea7aeda5a74a11c7b990e0f56ef23c99a83/src/ui/public/courier/data_source/_abstract.js#L295-L295

Bargs commented on src/ui/public/field_wildcard/field_wildcard.js:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  function fieldWildcardFilter(globs) {
+    const matcher = fieldWildcardMatcher(globs, config);
```
Looks like this config param is unnecessary

Bargs commented:
`test/output/.empty` can be deleted, it was deleted in master

Bargs commented:
@LeeDr could you review the added functional tests and `dump_data` json files in this PR? I'm not too familiar with elasticdump so I don't know if this looks right or not.

Bargs commented on src/ui/public/index_patterns/_index_pattern.js:
```diff
@@ -42,6 +43,7 @@ export default function IndexPatternFactory(Private, Notifier, config, kbnIndex,
     notExpandable: 'boolean',
     intervalName: 'string',
     fields: 'json',
+    fieldFilters: 'json',
```
sourceFilters

Bargs commented:
@scampi all done reviewing. I left a number of comments, but it's mostly small stuff, the functionality seems to be working well!

scampi commented:
@Bargs Addressed the problems you raised:
- renamed field filters to source filters everywhere, it was just laziness on my side...
- renamed the `_field_types.js` file to `_edit_sections.js` to better reflect what it is for
- I changed the layout of the source filtering edit to be a two-column layout:
 - for some dummy data, I have two source filters, `name*` and `nat*`.
 - In contrast to the previous design, now the "known fields" matching a pattern is displayed on the same row, instead of on the bottom and all together. I think this can simplify edit/debug the patterns when there are many.
 - the `rrr` pattern shows what happens if there are no matches for it.
![2016-08-14-171031_1176x711_scrot](https://cloud.githubusercontent.com/assets/795879/17651481/a414b0f6-625f-11e6-9773-58f8747a157d.png)

LeeDr commented:
@Bargs @scampi I'm not going to be able to get to this for another day or two.  If that's a problem we might need to find another reviewer.

scampi commented:
About the dumps, I took the ones already there, and added the `sourceFilters` field. Should be fine ;o)

Bargs commented:
@scampi digging the two column layout, I have a couple suggestions:

* This page looks very different than the other tabs (fields and scripted fields) but I don't think it needs to. I think we could use the same table directive those other tabs are using. It would have three columns, "filter", "matches", and "controls" (edit and delete button). Font and button size should match the other tabs as well. Instead of bullet points, the matches could be a comma delimited list. How does that sound?

* I noticed the placeholder text for the filter input isn't working.

* Why do the `.keyword` fields match? As far as I know, they don't actually get filtered anywhere because they're not part of the source?

scampi commented:
I changed the layout so that it looks more like the other sections:

- At the start:

![2016-08-28-225510_1071x711_scrot](https://cloud.githubusercontent.com/assets/795879/18037447/3e2ed582-6d7c-11e6-9a22-20c29b0180be.png)

- after adding few source filters

![2016-08-28-230056_1071x711_scrot](https://cloud.githubusercontent.com/assets/795879/18037454/52288bb4-6d7c-11e6-83a8-0c1c5e39d7b3.png)

- editing a filter

![2016-08-28-230122_1071x711_scrot](https://cloud.githubusercontent.com/assets/795879/18037457/717bbc52-6d7c-11e6-834c-bfe82d212f26.png)

- keep the source filters that match `ion`. The filtering is performed on the values of the `matches` column.

![2016-08-28-230241_1071x711_scrot](https://cloud.githubusercontent.com/assets/795879/18037460/8e708676-6d7c-11e6-8d0c-60274d81509b.png)

Bargs commented:
@scampi this looks awesome! ready for another review?

scampi commented:
Yes it is @Bargs , next time I'll comment if it is the case

Bargs commented on src/ui/public/field_wildcard/field_wildcard.js:
```diff
@@ -0,0 +1,35 @@
+  function fieldWildcardMatcher(globs) {
+    return function matcher(val) {
+      // do not test metaFields or keyword
+      if (metaFields.indexOf(val) !== -1 || endsWith(val, '.keyword')) {
```
You could use Lodash's endsWith instead: https://lodash.com/docs/3.10.1#endsWith

scampi replied:
good point

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/source_filters/source_filters.js:
```diff
@@ -0,0 +1,119 @@
+        this.placeHolder = 'source filter, accepts wildcards (e.g., `user*` to filter fields starting with \'user\')';
+
+        $scope.$watchMulti([ '[]indexPattern.sourceFilters', '$parent.fieldFilter' ], () => {
+          invoke(rowScopes, '$destroy');
```
Are you sure this is necessary? I would expect these scopes to get cleaned up when you dereference all the old row objects on line 60.

scampi replied:
to be honest, I took this from https://github.com/elastic/kibana/blob/master/src/core_plugins/kibana/public/management/sections/indices/_indexed_fields.js#L35 ;s

Bargs replied:
hmmm, in that case let's just leave it in. Might as well be consistent between sections

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/source_filters/source_filters.js:
```diff
@@ -0,0 +1,119 @@
+          {
+            title: 'matches',
+            sortable: false,
+            info: 'The known fields that match the filter'
```
Since our matching is imperfect, I'd add a little more explanation here:

"The source fields that match the filter. Multi-fields may show up as a false match but these filters only actually apply to fields in the original source document".

Bargs commented on src/core_plugins/kibana/public/management/sections/indices/source_filters/source_filters_controls.html:
```diff
@@ -0,0 +1,24 @@
+<div class="controls">
```
This file doesn't appear to be used anywhere, accidental duplicate of controls.html?

scampi commented:
@Bargs good catch on that bug! This is corrected in that last commit.
Ready for another round ;o)

Bargs commented:
LGTM! Awesome work @scampi, thanks for all your patience through this long review. Since things have changed quite a bit since the previous review I'm going to get one more team member to take a look before merging.

scampi commented:
great @Bargs ;o) ... Will you run the tests again ? I am not running them every time, I mostly lean on the automatic run of tests.

Bargs commented:
jenkins, test this

scampi commented:
🙏

scampi commented:
💰 where is that ninja emoji ? ;o)

epixa commented:
@scampi Can you update the title to be more descriptive (release notes worthy) and the description to include more comprehensive details about the change?  This will help other reviewers as well as make it easier to manage come release time.

epixa commented:
jenkins, test this

scampi commented:
I ran the tests and these are the errors messages:

- for '/elasticsearch/_msearch?timeout=0&ignore_unavailable=true&preference=1429577952339'

```js
{
  "type": "illegal_argument_exception",
  "reason": "request [/_msearch] contains unused params: [ignore_unavailable, preference, timeout]"
}
```

- for '/elasticsearch/_mget?timeout=0&ignore_unavailable=true&preference=1429574531063'

```js
{
  "type": "illegal_argument_exception",
  "reason": "request [/_mget] contains unused params: [ignore_unavailable, timeout]"
}
```

Do you know what this means ? This PR didn't change that part.

scampi commented:
@epixa Updated title + description

epixa commented:
If you merge the latest master into your branch, it should fix those tests.

> On Oct 6, 2016, at 4:51 AM, Stéphane Campinas <notifications@github.com> wrote:
> 
> I ran the tests and these are the errors messages:
> 
> for '/elasticsearch/_msearch?timeout=0&ignore_unavailable=true&preference=1429577952339'
> {
>   "type": "illegal_argument_exception",
>   "reason": "request [/_msearch] contains unused params: [ignore_unavailable, preference, timeout]"
> }
> for '/elasticsearch/_mget?timeout=0&ignore_unavailable=true&preference=1429574531063'
> {
>   "type": "illegal_argument_exception",
>   "reason": "request [/_mget] contains unused params: [ignore_unavailable, timeout]"
> }
> Do you know what this means ? This PR didn't change that part.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

scampi commented:
I got this error running UI tests:

```
Discover: [illegal_argument_exception] Deprecated field [exclude] used, expected [excludes] instead' to equal 'Discover: Saved Data Source "Query # 1
```

The commit 925641292991d60ca8c68a8fe7eabf5b7fa31e85 resolves that.

Browser tests failed with this strange error:

```
Chrome 53.0.2785 (Linux 0.0.0): Executed 2437 of 2550 SUCCESS (0 secs / 15.109 secs)
Chrome 53.0.2785 (Linux 0.0.0) ERROR
  Disconnectedundefined
Chrome 53.0.2785 (Linux 0.0.0): Executed 2437 of 2550 DISCONNECTED (1 min 59.603 secs / 15.109 sChrome 53.0.2785 (Linux 0.0.0): Executed 2437 of 2550 DISCONNECTED (1 min 59.603 secs / 15.109 secs)
```

epixa commented:
jenkins, test this

Bargs commented:
@scampi could you add a screenshot or two to the description as well? That'll help the next reviewer understand where to look.

scampi commented:
@Bargs I updated the description with some screenshots that show the major changes.

scampi commented:
@cjcenizal done

epixa commented:
jenkins, test this

spalger commented on test/functional/apps/discover/_source_filters.js:
```diff
@@ -0,0 +1,53 @@
+      //the refresh button to hide before clicking the share button
+      return PageObjects.common.sleep(1000);
+    })
+    .catch(PageObjects.common.createErrorHandler(this));
```
`PageObjects.common.createErrorHandler(this);` is no longer needed

scampi commented on src/core_plugins/kibana/public/management/sections/indices/_edit_sections.js:
```diff
@@ -0,0 +1,33 @@
+      {
+        title: 'source filters',
+        index: 'sourceFilters',
+        count: fieldCount.sourceFilters
```
The count is updated at https://github.com/elastic/kibana/pull/7402/files#diff-042b9dd3484133b320590d4492f96a4eR85

scampi commented:
@spalger addressed your comments in https://github.com/elastic/kibana/pull/7402/commits/0979aac74d9895adad0cfcb7e96af156f7289509

spalger commented:
jenkins, test this

spalger commented:
@Bargs can you take another look at this?

Bargs commented:
Just one last thought:

> Note that multi-fields will incorrectly appear as matches in the table below. These filters only actually apply to fields in the original source document, so matching multi-fields are not actually being filtered.

Now that we're putting this front and center, I think we should remove the [special treatment](https://github.com/elastic/kibana/pull/7402/files#diff-9b79cf7ceb0cbdae7ed42d0af0a438b3R13) of .raw and .keyword fields. Then at least our descriptive text and code will be consistent and hopefully less magical/confusing to users who are trying to understand how this works.

scampi commented:
@Bargs I removed the logic about raw and keyword fields.

Bargs commented:
LGTM. How about you @spalger ?

spalger commented:
big +1, LGTM!

Bargs commented:
🎉 🎉 🎉 🎉

epixa commented:
@scampi Thanks for sticking with this PR to the end.  This will help a lot of people.
